_posts/2015-07-20-third-party-integration.md

```diff
@@ -0,0 +1,85 @@
+
+However, with regards to third-party components we have to distinguish between third-party library and third-party web service.
+Third-party libraries are integrated directly into the code base and remain unchanged until consciously a new library version will be included.
+
+Working with third-party web services is more complex.
+Even if a third-party web service, other than a third-party library, is not directly expands the code base, one must never forget:
+By integrating a third-party service, this service will be part of the own software.
+The software will be expanded by functionality that is dependent on the availability of the third-party web service and acts as expected.
+This fact puts special demands on the developers.
+
+# The challenge
+
+Mostly one knows only the endpoint of the third-party API and only a more or less useful documentation is available.
+The problem is, that not only the web service itself, but also both, the endpoint and the documentation, can change every time.
+You will only notice the change, if something does not work anymore.
+Sometimes an SDK exists for the API that should facilitate working with it.
+But sometimes the SDK does not support the latest API version or the required programming language is not available.
+
+Often the documentation was only produced in the national language of the API developers making it impossible to understand the content.
+The result is that integration problems cannot be fixed by our developers in Barcelona if the documentation is e.g. only available in German.
+
+# The approach
+
+Ideally, we are in direct contact with the developers of the respective third-party to clarify any questions or concerns.
+Unfortunately, that is not always the case and we have to find it out by ourselves with the trial-and-error method or reverse engineering.
+
+When working with a SOAP API, the WSDL-file that describes the service, provides many helpful information.
+Using a tool such as SoapUI, such a file can be analysed easily.
+Even test requests can be sent via the SoapUI.
+
+For such tests, the third-party provides special test accounts.
+Unfortunately, these are sometimes only valid for a limited period of time.
+This is only then noticed, if the account is required again, e.g. if after some time problems with an integration arise that have to be fixed on short notice.
+
+To test our integrations we are writing Unit tests as well as integration and Selenium tests.
+By carrying out these test we are alerted to these problems already at an early stage.
+
+Some companies even operate special test platforms with a separate endpoint.
+What sounds like a very depeloper-friendly offer, has a major disadvantage: doing the tests on one system and using a different system for production.
+Ideally, the test system should act exactly as the live-system, but that is not always the case.
+This is especially annoying if after the rollout customers report problems that did not arise on the test platform.
```

Comment:
Here i am just describing the problems which probably every company has, which is integrating third party webservices. Since the problems I described here are no explicit bugs and I am not pointing directly on certain companies, I wouldn't consider this as blaming.